# Importing all necessary dependencies

In [1]:
import tensorflow as tf
import numpy as np
import pandas as pd
from sklearn import cross_validation
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn import preprocessing
from sklearn.cross_validation import train_test_split
from random import sample

/home/sakshi/anaconda2/lib/python2.7/site-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


# Getting the data

In [2]:
data = pd.read_csv("/home/sakshi/Desktop/processed.cleveland.csv", names = ['age','sex','cp','BP','chol','fbs','ecg','maxhr','eiang','eist','slope','vessels','thal','diagnosis'], na_values=["?"])

In [3]:
data

age  sex   cp     BP   chol  fbs  ecg  maxhr  eiang  eist  slope  \
0    63.0  1.0  1.0  145.0  233.0  1.0  2.0  150.0    0.0   2.3    3.0   
1    67.0  1.0  4.0  160.0  286.0  0.0  2.0  108.0    1.0   1.5    2.0   
2    67.0  1.0  4.0  120.0  229.0  0.0  2.0  129.0    1.0   2.6    2.0   
3    37.0  1.0  3.0  130.0  250.0  0.0  0.0  187.0    0.0   3.5    3.0   
4    41.0  0.0  2.0  130.0  204.0  0.0  2.0  172.0    0.0   1.4    1.0   
5    56.0  1.0  2.0  120.0  236.0  0.0  0.0  178.0    0.0   0.8    1.0   
6    62.0  0.0  4.0  140.0  268.0  0.0  2.0  160.0    0.0   3.6    3.0   
7    57.0  0.0  4.0  120.0  354.0  0.0  0.0  163.0    1.0   0.6    1.0   
8    63.0  1.0  4.0  130.0  254.0  0.0  2.0  147.0    0.0   1.4    2.0   
9    53.0  1.0  4.0  140.0  203.0  1.0  2.0  155.0    1.0   3.1    3.0   
10   57.0  1.0  4.0  140.0  192.0  0.0  0.0  148.0    0.0   0.4    2.0   
11   56.0  0.0  2.0  140.0  294.0  0.0  2.0  153.0    0.0   1.3    2.0   
12   56.0  1.0  3.0  130.0  256.0  1.0  2.0  142.0    1.0   0.6    2.0   
13   44.0  1.0  2.0  120.0  263.0  0.0  0.0  173.0    0.0   0.0    1.0   
14   52.0  1.0  3.0  172.0  199.0  1.0  0.0  162.0    0.0   0.5    1.0   
15   57.0  1.0  3.0  150.0  168.0  0.0  0.0  174.0    0.0   1.6    1.0   
16   48.0  1.0  2.0  110.0  229.0  0.0  0.0  168.0    0.0   1.0    3.0   
17   54.0  1.0  4.0  140.0  239.0  0.0  0.0  160.0    0.0   1.2    1.0   
18   48.0  0.0  3.0  130.0  275.0  0.0  0.0  139.0    0.0   0.2    1.0   
19   49.0  1.0  2.0  130.0  266.0  0.0  0.0  171.0    0.0   0.6    1.0   
20   64.0  1.0  1.0  110.0  211.0  0.0  2.0  144.0    1.0   1.8    2.0   
21   58.0  0.0  1.0  150.0  283.0  1.0  2.0  162.0    0.0   1.0    1.0   
22   58.0  1.0  2.0  120.0  284.0  0.0  2.0  160.0    0.0   1.8    2.0   
23   58.0  1.0  3.0  132.0  224.0  0.0  2.0  173.0    0.0   3.2    1.0   
24   60.0  1.0  4.0  130.0  206.0  0.0  2.0  132.0    1.0   2.4    2.0   
25   50.0  0.0  3.0  120.0  219.0  0.0  0.0  158.0    0.0   1.6    2.0   
26   58.0  0.0  3.0  120.0  340.0  0.0  0.0  172.0    0.0   0.0    1.0   
27   66.0  0.0  1.0  150.0  226.0  0.0  0.0  114.0    0.0   2.6    3.0   
28   43.0  1.0  4.0  150.0  247.0  0.0  0.0  171.0    0.0   1.5    1.0   
29   40.0  1.0  4.0  110.0  167.0  0.0  2.0  114.0    1.0   2.0    2.0   
..    ...  ...  ...    ...    ...  ...  ...    ...    ...   ...    ...   
273  71.0  0.0  4.0  112.0  149.0  0.0  0.0  125.0    0.0   1.6    2.0   
274  59.0  1.0  1.0  134.0  204.0  0.0  0.0  162.0    0.0   0.8    1.0   
275  64.0  1.0  1.0  170.0  227.0  0.0  2.0  155.0    0.0   0.6    2.0   
276  66.0  0.0  3.0  146.0  278.0  0.0  2.0  152.0    0.0   0.0    2.0   
277  39.0  0.0  3.0  138.0  220.0  0.0  0.0  152.0    0.0   0.0    2.0   
278  57.0  1.0  2.0  154.0  232.0  0.0  2.0  164.0    0.0   0.0    1.0   
279  58.0  0.0  4.0  130.0  197.0  0.0  0.0  131.0    0.0   0.6    2.0   
280  57.0  1.0  4.0  110.0  335.0  0.0  0.0  143.0    1.0   3.0    2.0   
281  47.0  1.0  3.0  130.0  253.0  0.0  0.0  179.0    0.0   0.0    1.0   
282  55.0  0.0  4.0  128.0  205.0  0.0  1.0  130.0    1.0   2.0    2.0   
283  35.0  1.0  2.0  122.0  192.0  0.0  0.0  174.0    0.0   0.0    1.0   
284  61.0  1.0  4.0  148.0  203.0  0.0  0.0  161.0    0.0   0.0    1.0   
285  58.0  1.0  4.0  114.0  318.0  0.0  1.0  140.0    0.0   4.4    3.0   
286  58.0  0.0  4.0  170.0  225.0  1.0  2.0  146.0    1.0   2.8    2.0   
287  58.0  1.0  2.0  125.0  220.0  0.0  0.0  144.0    0.0   0.4    2.0   
288  56.0  1.0  2.0  130.0  221.0  0.0  2.0  163.0    0.0   0.0    1.0   
289  56.0  1.0  2.0  120.0  240.0  0.0  0.0  169.0    0.0   0.0    3.0   
290  67.0  1.0  3.0  152.0  212.0  0.0  2.0  150.0    0.0   0.8    2.0   
291  55.0  0.0  2.0  132.0  342.0  0.0  0.0  166.0    0.0   1.2    1.0   
292  44.0  1.0  4.0  120.0  169.0  0.0  0.0  144.0    1.0   2.8    3.0   
293  63.0  1.0  4.0  140.0  187.0  0.0  2.0  144.0    1.0   4.0    1.0   
294  63.0  0.0  4.0  124.0  197.0  0.0  0.0  136.0    1.0   0.0    2.0   
295  41.0 

# Checking for the count among all columns

In [4]:
data.count()

age          303
sex          303
cp           303
BP           303
chol         303
fbs          303
ecg          303
maxhr        303
eiang        303
eist         303
slope        303
vessels      299
thal         301
diagnosis    303
dtype: int64

We can clearly see that 'vessels' and 'thal' columns have NaN values

# Getting the rows with Null values in 'vessels'

In [5]:
data[data.vessels.isnull()]

age  sex   cp     BP   chol  fbs  ecg  maxhr  eiang  eist  slope  \
166  52.0  1.0  3.0  138.0  223.0  0.0  0.0  169.0    0.0   0.0    1.0   
192  43.0  1.0  4.0  132.0  247.0  1.0  2.0  143.0    1.0   0.1    2.0   
287  58.0  1.0  2.0  125.0  220.0  0.0  0.0  144.0    0.0   0.4    2.0   
302  38.0  1.0  3.0  138.0  175.0  0.0  0.0  173.0    0.0   0.0    1.0   

     vessels  thal  diagnosis  
166      NaN   3.0          0  
192      NaN   7.0          1  
287      NaN   7.0          0  
302      NaN   3.0          0

# Getting the rows with Null values in 'thal'

In [6]:
data[data.thal.isnull()]

age  sex   cp     BP   chol  fbs  ecg  maxhr  eiang  eist  slope  \
87   53.0  0.0  3.0  128.0  216.0  0.0  2.0  115.0    0.0   0.0    1.0   
266  52.0  1.0  4.0  128.0  204.0  1.0  0.0  156.0    1.0   1.0    2.0   

     vessels  thal  diagnosis  
87       0.0   NaN          0  
266      0.0   NaN          2

Now that we know that we have Null values in thal and vessels columns so, we can replace these values with their mean values in their respective classes

# Getting the classes with missing values

In [7]:
data.groupby('diagnosis').count()

age  sex   cp   BP  chol  fbs  ecg  maxhr  eiang  eist  slope  \
diagnosis                                                                  
0          164  164  164  164   164  164  164    164    164   164    164   
1           55   55   55   55    55   55   55     55     55    55     55   
2           36   36   36   36    36   36   36     36     36    36     36   
3           35   35   35   35    35   35   35     35     35    35     35   
4           13   13   13   13    13   13   13     13     13    13     13   

           vessels  thal  
diagnosis                 
0              161   163  
1               54    55  
2               36    35  
3               35    35  
4               13    13

We can see we have 3 missing values of 'vessels' in class0 and 1 in class1, similarly 'thal' has 1 missing value in class0 and 1 in class2. We will replace them with their means in respective classes.

In [8]:
# Considering class0, class1, class2 only, as only these classes have NaN values
class0 = data[data.diagnosis==0]
class1 = data[data.diagnosis==1]
class2 = data[data.diagnosis==2]

# Handling NaN values in 'vessels' columns

In [9]:
class0Mean_vessels = data[data.diagnosis==0].vessels.mean()
class0Mean_vessels

0.2732919254658385

In [10]:
class1Mean_vessels = data[data.diagnosis==1].vessels.mean()
class1Mean_vessels

0.7407407407407407

We see for class0 we have mean as 0.268293 and for class1 it is 0.727273, we can get their rounded off values as vessels stores the record for number of major vessels colored by fluoroscopy whose value lies between [0, 3]

In [11]:
vesseslsVal0_vessels = round(class0Mean_vessels)
vesseslsVal0_vessels

0.0

In [12]:
vesseslsVal1_vessels = round(class1Mean_vessels)
vesseslsVal1_vessels

1.0

We get rounded values as 0 and 1 for class0 and class1 respectively, now we got to update our dataset with these values

# Handling NaN values in 'thal' columns

In [13]:
# Updating class0 and class1 with the new rounded off mean calculated
class0.vessels.fillna(vesseslsVal0_vessels, axis=0, inplace=True)
class1.vessels.fillna(vesseslsVal1_vessels, axis=0, inplace=True)

/home/sakshi/anaconda2/lib/python2.7/site-packages/pandas/core/generic.py:3549: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


In [14]:
class0Mean_thal = data[data.diagnosis==0].thal.mean()
class0Mean_thal

3.7975460122699385

In [15]:
class2Mean_thal = data[data.diagnosis==2].thal.mean()
class2Mean_thal

6.0285714285714285

We see for class0 we have mean as 3.797546 and for class2 it is 6.028571, we can get their rounded off values, as variable "thal" refers to a Thalium Stress Test and is represented by the integers 3, 6, and 7(3=Normal; 6=Fixed Defect; and 7=Reversible Defect).

In [16]:
thalVal0_thal = int(class0Mean_thal)
thalVal0_thal

3

In [17]:
thalVal2_thal = round(class2Mean_thal)
thalVal2_thal

6.0

We get rounded values as 3 and 6 for class0 and class2 respectively, now we got to update our dataset with these values

In [18]:
# Updating class0 and class1 with the new rounded off mean calculated
class0.thal.fillna(thalVal0_thal, axis=0, inplace=True)
class2.thal.fillna(thalVal2_thal, axis=0, inplace=True)

# Removing classes with NaN values from dataset

In [19]:
data = data[data.diagnosis != 0] 
data = data[data.diagnosis != 1] 
data = data[data.diagnosis != 2]

# Changing the target values that are 2, 3 or 4 to 1
data.diagnosis[data.diagnosis==2] = 1
data.diagnosis[data.diagnosis==3] = 1
data.diagnosis[data.diagnosis==4] = 1

/home/sakshi/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys
/home/sakshi/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


# Adding processed classes values to our dataset

In [20]:
data = pd.concat([class0, class1, class2, data])

In [21]:
data

age  sex   cp     BP   chol  fbs  ecg  maxhr  eiang  eist  slope  \
0    63.0  1.0  1.0  145.0  233.0  1.0  2.0  150.0    0.0   2.3    3.0   
3    37.0  1.0  3.0  130.0  250.0  0.0  0.0  187.0    0.0   3.5    3.0   
4    41.0  0.0  2.0  130.0  204.0  0.0  2.0  172.0    0.0   1.4    1.0   
5    56.0  1.0  2.0  120.0  236.0  0.0  0.0  178.0    0.0   0.8    1.0   
7    57.0  0.0  4.0  120.0  354.0  0.0  0.0  163.0    1.0   0.6    1.0   
10   57.0  1.0  4.0  140.0  192.0  0.0  0.0  148.0    0.0   0.4    2.0   
11   56.0  0.0  2.0  140.0  294.0  0.0  2.0  153.0    0.0   1.3    2.0   
13   44.0  1.0  2.0  120.0  263.0  0.0  0.0  173.0    0.0   0.0    1.0   
14   52.0  1.0  3.0  172.0  199.0  1.0  0.0  162.0    0.0   0.5    1.0   
15   57.0  1.0  3.0  150.0  168.0  0.0  0.0  174.0    0.0   1.6    1.0   
17   54.0  1.0  4.0  140.0  239.0  0.0  0.0  160.0    0.0   1.2    1.0   
18   48.0  0.0  3.0  130.0  275.0  0.0  0.0  139.0    0.0   0.2    1.0   
19   49.0  1.0  2.0  130.0  266.0  0.0  0.0  171.0    0.0   0.6    1.0   
20   64.0  1.0  1.0  110.0  211.0  0.0  2.0  144.0    1.0   1.8    2.0   
21   58.0  0.0  1.0  150.0  283.0  1.0  2.0  162.0    0.0   1.0    1.0   
25   50.0  0.0  3.0  120.0  219.0  0.0  0.0  158.0    0.0   1.6    2.0   
26   58.0  0.0  3.0  120.0  340.0  0.0  0.0  172.0    0.0   0.0    1.0   
27   66.0  0.0  1.0  150.0  226.0  0.0  0.0  114.0    0.0   2.6    3.0   
28   43.0  1.0  4.0  150.0  247.0  0.0  0.0  171.0    0.0   1.5    1.0   
30   69.0  0.0  1.0  140.0  239.0  0.0  0.0  151.0    0.0   1.8    1.0   
33   59.0  1.0  4.0  135.0  234.0  0.0  0.0  161.0    0.0   0.5    2.0   
34   44.0  1.0  3.0  130.0  233.0  0.0  0.0  179.0    1.0   0.4    1.0   
35   42.0  1.0  4.0  140.0  226.0  0.0  0.0  178.0    0.0   0.0    1.0   
39   61.0  1.0  3.0  150.0  243.0  1.0  0.0  137.0    1.0   1.0    2.0   
41   40.0  1.0  1.0  140.0  199.0  0.0  0.0  178.0    1.0   1.4    1.0   
42   71.0  0.0  2.0  160.0  302.0  0.0  0.0  162.0    0.0   0.4    1.0   
43   59.0  1.0  3.0  150.0  212.0  1.0  0.0  157.0    0.0   1.6    1.0   
46   51.0  1.0  3.0  110.0  175.0  0.0  0.0  123.0    0.0   0.6    1.0   
48   65.0  0.0  3.0  140.0  417.0  1.0  2.0  157.0    0.0   0.8    1.0   
49   53.0  1.0  3.0  130.0  197.0  1.0  2.0  152.0    0.0   1.2    3.0   
..    ...  ...  ...    ...    ...  ...  ...    ...    ...   ...    ...   
121  63.0  0.0  4.0  150.0  407.0  0.0  2.0  154.0    0.0   4.0    2.0   
123  55.0  1.0  4.0  140.0  217.0  0.0  0.0  111.0    1.0   5.6    3.0   
126  56.0  0.0  4.0  200.0  288.0  1.0  2.0  133.0    1.0   4.0    3.0   
127  54.0  1.0  4.0  110.0  239.0  0.0  0.0  126.0    1.0   2.8    2.0   
136  70.0  1.0  4.0  145.0  174.0  0.0  0.0  125.0    1.0   2.6    3.0   
137  62.0  1.0  2.0  120.0  281.0  0.0  2.0  103.0    0.0   1.4    2.0   
146  57.0  1.0  4.0  165.0  289.0  1.0  2.0  124.0    0.0   1.0    2.0   
153  55.0  1.0  4.0  160.0  289.0  0.0  2.0  145.0    1.0   0.8    2.0   
154  64.0  1.0  4.0  120.0  246.0  0.0  2.0   96.0    1.0   2.2    3.0   
161  77.0  1.0  4.0  125.0  304.0  0.0  2.0  162.0    1.0   0.0    1.0   
170  70.0  1.0  3.0  160.0  269.0  0.0  0.0  112.0    1.0   2.9    2.0   
174  64.0  1.0  4.0  145.0  212.0  0.0  2.0  132.0    0.0   2.0    2.0   
180  48.0  1.0  4.0  124.0  274.0  0.0  2.0  166.0    0.0   0.5    2.0   
191  51.0  1.0  4.0  140.0  298.0  0.0  0.0  122.0    1.0   4.2    2.0   
195  67.0  1.0  4.0  100.0  299.0  0.0  2.0  125.0    1.0   0.9    2.0   
205  45.0  1.0  4.0  142.0  309.0  0.0  2.0  147.0    1.0   0.0    2.0   
206  58.0  1.0  4.0  128.0  259.0  0.0  2.0  130.0    1.0   3.0    2.0   
207  50.0  1.0  4.0  144.0  200.0  0.0  2.0  126.0    1.0   0.9    2.0   
211  38.0  1.0  1.0  120.0  231.0  0.0  0.0  182.0    1.0   3.8    2.0   
213  66.0  0.0  4.0  178.0  228.0  1.0  0.0  165.0    1.0   1.0    2.0   
223  53.0  1.0  4.0  123.0  282.0  0.0  0.0   95.0    1.0   2.0    2.0   
228  54.0  1.0  4.0  110.0  206.0  0.0  2.0  108.0    1.0   0.0    2.0   
235  54.0 

In [22]:
# Sorting the data by it's index
data = data.sort_index()

# Chnaging the categorical values with their labels

In [23]:
data.loc[data.cp == 1, 'cp'] = 'typical_angina'
data.loc[data.cp == 2, 'cp'] = 'atypical_angina'
data.loc[data.cp == 3, 'cp'] = 'non_anginal_pain'
data.loc[data.cp == 4, 'cp'] = 'asymptomatic'
data.loc[data.ecg == 0, 'ecg'] = 'normal'
data.loc[data.ecg == 1, 'ecg'] = 'STT_wave_abnormality'
data.loc[data.ecg == 2, 'ecg'] = 'left_ventricular_hypertrophy'
data.loc[data.eiang == 0, 'eiang'] = 'no'
data.loc[data.eiang == 1, 'eiang'] = 'yes'
data.loc[data.slope == 1, 'slope'] = 'upsloping'
data.loc[data.slope == 2, 'slope'] = 'flat'
data.loc[data.slope == 3, 'slope'] = 'downsloping'
data.loc[data.thal == 3, 'thal'] = 'normal'
data.loc[data.thal == 6, 'thal'] = 'fixed_defect'
data.loc[data.thal == 7, 'thal'] = 'reversable_defect'

# Getting the dummy values

In [24]:
data = pd.get_dummies(data)
data.head().T

0      1      2      3      4
age                                63.0   67.0   67.0   37.0   41.0
sex                                 1.0    1.0    1.0    1.0    0.0
BP                                145.0  160.0  120.0  130.0  130.0
chol                              233.0  286.0  229.0  250.0  204.0
fbs                                 1.0    0.0    0.0    0.0    0.0
maxhr                             150.0  108.0  129.0  187.0  172.0
eist                                2.3    1.5    2.6    3.5    1.4
vessels                             0.0    3.0    2.0    0.0    0.0
diagnosis                           0.0    2.0    1.0    0.0    0.0
cp_asymptomatic                     0.0    1.0    1.0    0.0    0.0
cp_atypical_angina                  0.0    0.0    0.0    0.0    1.0
cp_non_anginal_pain                 0.0    0.0    0.0    1.0    0.0
cp_typical_angina                   1.0    0.0    0.0    0.0    0.0
ecg_STT_wave_abnormality            0.0    0.0    0.0    0.0    0.0
ecg_left_ventricular_hypertrophy    1.0    1.0    1.0    0.0    1.0
ecg_normal                          0.0    0.0    0.0    1.0    0.0
eiang_no                            1.0    0.0    0.0    1.0    1.0
eiang_yes                           0.0    1.0    1.0    0.0    0.0
slope_downsloping                   1.0    0.0    0.0    1.0    0.0
slope_flat                          0.0    1.0    1.0    0.0    0.0
slope_upsloping                     0.0    0.0    0.0    0.0    1.0
thal_fixed_defect                   1.0    0.0    0.0    0.0    0.0
thal_normal                         0.0    1.0    0.0    1.0    1.0
thal_reversable_defect              0.0    0.0    1.0    0.0    0.0

In [25]:
x_data = data[['age','sex','cp_asymptomatic','cp_atypical_angina','cp_non_anginal_pain','cp_typical_angina','BP','chol','fbs','ecg_STT_wave_abnormality','ecg_left_ventricular_hypertrophy','ecg_normal','maxhr','eiang_no','eiang_yes','eist','slope_downsloping','slope_flat','slope_upsloping','vessels','thal_fixed_defect','thal_normal','thal_reversable_defect']]
y_data = data['diagnosis']
x_data.shape

(303, 23)

# Dividing data into training and testing portions

In [26]:
x_train, x_test, y_train, y_test = cross_validation.train_test_split(x_data, y_data, test_size=0.25, random_state=42)
labels_train = np.c_[y_train, 1 - y_train]
labels_train = np.float32(labels_train)

labels_test = np.c_[y_test, 1 - y_test]
labels_test = np.float32(labels_test)
x_test.shape, labels_train.shape

((76, 23), (227, 2))

# Normalizing the training and testing data  

In [27]:
from sklearn import preprocessing
#Normalizing the data
min_max_scaler = preprocessing.MinMaxScaler().fit(x_train)
x_train = min_max_scaler.transform(x_train)
x_test = min_max_scaler.transform(x_test)

# Model

In [28]:
data_train = np.c_[x_train, labels_train]
import tensorflow as tf

# Reset Graph
tf.reset_default_graph()

num_features = 23
log_path = 'logs/'

# Hyper Parameters
learning_rate   = 0.001
training_epochs = 10000
batch_size      = 100
display_step    = 100

# Network Parameters
n_hidden_1 = 23 # 1st layer number of neurons
n_hidden_2 = 30 # 2nd layer number of neurons
n_hidden_3 = 40 # 3nd layer number of neurons
n_input    = 23 # data input
n_classes  = 2  # total classes 

# Placeholders
x = tf.placeholder("float", [None, n_input], name='x')
y = tf.placeholder("float", [None, n_classes], name='y')

x.shape, y.shape

(TensorShape([Dimension(None), Dimension(23)]),
 TensorShape([Dimension(None), Dimension(2)]))

In [29]:
def variable_summaries(var):
    # Attach a scope for the summary
    with tf.name_scope("summaries"):
        mean = tf.reduce_mean(var)
        tf.summary.scalar('mean', mean)
        
        with tf.name_scope('stddev'):
            stddev = tf.sqrt(tf.reduce_mean(tf.square(var - mean)))
        
        tf.summary.scalar('stddev', stddev)
        tf.summary.scalar('max', tf.reduce_max(var))
        tf.summary.scalar('min', tf.reduce_min(var))
        tf.summary.histogram('histogram', var)        

In [30]:
def nn_layer(input_tensor, input_dim, output_dim, layer_name, act=tf.nn.relu):
     
    """Reusable code for making a simple neural net layer.

    It does a matrix multiply, bias add, and then uses relu to nonlinearize.
    It also sets up name scoping so that the resultant graph is easy to read,
    and adds a number of summary ops.
    """
    # Adding a name scope ensures logical grouping of the layers in the graph.
    with tf.name_scope(layer_name):
        # This Variable will hold the state of the weights for the layer
        with tf.name_scope('weights'):
            weights = tf.Variable(tf.truncated_normal([input_dim, output_dim], stddev=0.1), name='weights')
            variable_summaries(weights)
        with tf.name_scope('biases'):
            biases = tf.Variable(tf.constant(0.1, shape=([output_dim])), name='biases')
            variable_summaries(biases)
        with tf.name_scope('Wx_plus_b'):
            preacts = tf.add(tf.matmul(input_tensor, weights), biases)
            tf.summary.histogram('pre_activations', preacts)
        activations = act(preacts, name="activation")
        tf.summary.histogram('activations', activations)
    
    return activations

In [31]:
hidden1 = nn_layer(x, num_features, n_hidden_1, "layer1")
hidden2 = nn_layer(hidden1, n_hidden_1, n_hidden_2, "layer2")
hidden3 = nn_layer(hidden2, n_hidden_2, n_hidden_3, "layer3")

# Final layer - Use tf.identity to make sure there are no activations
logits = nn_layer(hidden3, n_hidden_3, n_classes, "softmax", act=tf.identity)

In [32]:
# Compute and minimize the cross entropy

cross_entropy= tf.nn.softmax_cross_entropy_with_logits(labels = y, logits = logits, name = 'cross_entropy')
cross_entropy = tf.reduce_mean(cross_entropy)
tf.summary.scalar('cross_entropy', cross_entropy)

<tf.Tensor 'cross_entropy_1:0' shape=() dtype=string>

# Optimizer

In [33]:
train_step = tf.train.AdamOptimizer(learning_rate = learning_rate).minimize(cross_entropy)

# Calculating accuracy of our model 

In [34]:
correct_prediction = tf.equal(tf.argmax(logits, 1), tf.argmax(y, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32)) 
tf.summary.scalar("accuracy", accuracy)

<tf.Tensor 'accuracy:0' shape=() dtype=string>

In [35]:
# Merge all the summaries
summaries = tf.summary.merge_all()

# Initializing the variables
init = tf.global_variables_initializer()
total_batch = 5
sess = tf.InteractiveSession()
sess.run(init)

writer = tf.summary.FileWriter(log_path, sess.graph)

# Training cycle
for epoch in range(training_epochs):
    # Loop over all batches
    for batch in range(total_batch):
        samples = sample(range(0,data_train.shape[0]), batch_size)
        batch_x = data_train[samples,:23] 
        batch_y = data_train[samples,23:]
            
        curr_loss, cur_accuracy, _, summary = sess.run([cross_entropy, accuracy, train_step, summaries], feed_dict={x: batch_x, y: batch_y})
        writer.add_summary(_, batch)
            
    # Display logs per epoch step
    if epoch % (display_step * 10) == 0:
        writer.add_summary(summary, epoch * total_batch + batch)
        # Print the loss
        print("Epoch: %04d/%d. Current loss: %.8f. Train Accuracy: %.3f"
                      %(epoch, training_epochs, curr_loss, cur_accuracy))            
    
# Test the session
test_accuracy, test_predictions = sess.run([accuracy, tf.argmax(logits, 1)], feed_dict={ x: x_test, y: labels_test})
        
Accuracy_val = round(test_accuracy*100, 2)
print("Test Accuracy: %.3f" % Accuracy_val)

Epoch: 0000/10000. Current loss: 0.68598694. Train Accuracy: 0.450
Epoch: 1000/10000. Current loss: -14522689.00000000. Train Accuracy: 0.830
Epoch: 2000/10000. Current loss: -90691144.00000000. Train Accuracy: 0.800
Epoch: 3000/10000. Current loss: -352977056.00000000. Train Accuracy: 0.870
Epoch: 4000/10000. Current loss: -698666176.00000000. Train Accuracy: 0.770
Epoch: 5000/10000. Current loss: -2379722496.00000000. Train Accuracy: 0.840
Epoch: 6000/10000. Current loss: -6069454336.00000000. Train Accuracy: 0.830
Epoch: 7000/10000. Current loss: -6660302848.00000000. Train Accuracy: 0.840
Epoch: 8000/10000. Current loss: -7368500736.00000000. Train Accuracy: 0.810
Epoch: 9000/10000. Current loss: -12113653760.00000000. Train Accuracy: 0.780
Test Accuracy: 88.160
